File: c/sdk/data/other/Effect.c
#include "Effect.h"

const char* toStringEffectType(EffectType type) {
    switch (type) {
        case IMMEDIATE_SHOW:       return "IMMEDIATE_SHOW";
        case LEFT_PARALLEL_MOVE:    return "LEFT_PARALLEL_MOVE";
        case RIGHT_PARALLEL_MOVE:   return "RIGHT_PARALLEL_MOVE";
        case UP_PARALLEL_MOVE:      return "UP_PARALLEL_MOVE";
        case DOWN_PARALLEL_MOVE:    return "DOWN_PARALLEL_MOVE";
        case LEFT_COVER:            return "LEFT_COVER";
        case RIGHT_COVER:           return "RIGHT_COVER";
        case UP_COVER:              return "UP_COVER";
        case DOWN_COVER:            return "DOWN_COVER";
        case LEFT_TOP_COVER:        return "LEFT_TOP_COVER";
        case LEFT_BOTTOM_COVER:     return "LEFT_BOTTOM_COVER";
        case RIGHT_TOP_COVER:       return "RIGHT_TOP_COVER";
        case RIGHT_BOTTOM_COVER:    return "RIGHT_BOTTOM_COVER";
        case HORIZONTAL_DIVIDE:     return "HORIZONTAL_DIVIDE";
        case VERTICAL_DIVIDE:       return "VERTICAL_DIVIDE";
        case HORIZONTAL_CLOSE:      return "HORIZONTAL_CLOSE";
        case VERTICAL_CLOSE:        return "VERTICAL_CLOSE";
        case FADE:                  return "FADE";
        case HORIZONTAL_SHUTTER:    return "HORIZONTAL_SHUTTER";
        case VERTICAL_SHUTTER:      return "VERTICAL_SHUTTER";
        case NOT_CLEAR_AREA:        return "NOT_CLEAR_AREA";
        case LEFT_SERIES_MOVE:      return "LEFT_SERIES_MOVE";
        case RIGHT_SERIES_MOVE:     return "RIGHT_SERIES_MOVE";
        case UP_SERIES_MOVE:        return "UP_SERIES_MOVE";
        case DOWN_SERIES_MOVE:      return "DOWN_SERIES_MOVE";
        case RANDOM:                return "RANDOM";
        case HT_LEFT_SERIES_MOVE:   return "HT_LEFT_SERIES_MOVE";
        case HT_RIGHT_SERIES_MOVE:  return "HT_RIGHT_SERIES_MOVE";
        case HT_UP_SERIES_MOVE:     return "HT_UP_SERIES_MOVE";
        case HT_DOWN_SERIES_MOVE:   return "HT_DOWN_SERIES_MOVE";
        case FLICKEREFFECT:         return "FLICKEREFFECT";
        default: return "UNKNOWN";
    }
}

Effect* createEffect() {
    Effect* effect = (Effect*)hdmalloc(sizeof(Effect));
    effect->effectType = IMMEDIATE_SHOW;
    effect->speed = 5;
    effect->hold = 5000;
    return effect;
}

Effect* createWithTypeEffect(EffectType type) {
    Effect* effect = createEffect();
    effect->effectType = type;
    return effect;
}

Effect* createWithTypeAndSpeedEffect(EffectType type, int speed) {
    Effect* effect = createWithTypeEffect(type);
    effect->speed = speed;
    return effect;
}

Effect* createFullEffect(EffectType type, int speed, int hold) {
    Effect* effect = createWithTypeAndSpeedEffect(type, speed);
    effect->hold = hold;
    return effect;
}

Effect* copyEffect(const Effect* other) {
    return createFullEffect(other->effectType, other->speed, other->hold);
}

void destroyEffect(Effect* effect) {
    hdfree(effect);
}

EffectType getTypeEffect(const Effect* effect) {
    return effect ? effect->effectType : IMMEDIATE_SHOW;
}

void setTypeEffect(Effect* effect, EffectType type) {
    if (effect) {
        effect->effectType = type;
    }
}

int getTypeValueEffect(const Effect* effect) {
    return effect ? (int)effect->effectType : 0;
}

void setTypeValueEffect(Effect* effect, int type) {
    if (effect) {
        effect->effectType = (EffectType)type;
    }
}

int getSpeedEffect(const Effect* effect) {
    return effect ? effect->speed : 5;
}

void setSpeedEffect(Effect* effect, int speed) {
    if (effect) {
        effect->speed = speed;
    }
}

int getHoldEffect(const Effect* effect) {
    return effect ? effect->hold : 5000;
}

void setHoldEffect(Effect* effect, int hold) {
    if (effect) {
        effect->hold = hold;
    }
}

cJSON* toJsonEffect(const Effect* effect) {
    if (!effect) {
        return NULL;
    }
    cJSON* json = cJSON_CreateObject();
    cJSON_AddNumberToObject(json, "type", (int)effect->effectType);
    cJSON_AddNumberToObject(json, "speed", effect->speed);
    cJSON_AddNumberToObject(json, "hold", effect->hold);
    return json;
}

Effect* fromJsonEffect(const cJSON* json) {
    if (!json) {
        return NULL;
    }
    cJSON* typeJson = cJSON_GetObjectItemCaseSensitive(json, "type");
    cJSON* speedJson = cJSON_GetObjectItemCaseSensitive(json, "speed");
    cJSON* holdJson = cJSON_GetObjectItemCaseSensitive(json, "hold");

    Effect* effect = createEffect();
    if (typeJson) {
        effect->effectType = (EffectType)typeJson->valueint;
    }
    if (speedJson) {
        effect->speed = speedJson->valueint;
    }
    if (holdJson) {
        effect->hold = holdJson->valueint;
    }

    return effect;
}

File: python/sdk/data/other/Effect.py
from enum import IntEnum
import json

class Effect:
    class EffectType(IntEnum):
        IMMEDIATE_SHOW = 0         # Direct display
        LEFT_PARALLEL_MOVE = 1     # Left parallel move
        RIGHT_PARALLEL_MOVE = 2    # Right parallel move
        UP_PARALLEL_MOVE = 3       # Up parallel move
        DOWN_PARALLEL_MOVE = 4     # Down parallel move
        LEFT_COVER = 5             # Left cover
        RIGHT_COVER = 6            # Right cover
        UP_COVER = 7               # Up cover
        DOWN_COVER = 8             # Down cover
        LEFT_TOP_COVER = 9         # Top-left cover
        LEFT_BOTTOM_COVER = 10     # Bottom-left cover
        RIGHT_TOP_COVER = 11       # Top-right cover
        RIGHT_BOTTOM_COVER = 12    # Bottom-right cover
        HORIZONTAL_DIVIDE = 13     # Horizontal split
        VERTICAL_DIVIDE = 14       # Vertical split
        HORIZONTAL_CLOSE = 15      # Horizontal close
        VERTICAL_CLOSE = 16        # Vertical close
        FADE = 17                  # Fade in/out
        HORIZONTAL_SHUTTER = 18    # Horizontal blinds
        VERTICAL_SHUTTER = 19      # Vertical blinds
        NOT_CLEAR_AREA = 20        # Don't clear area
        LEFT_SERIES_MOVE = 21      # Continuous left move
        RIGHT_SERIES_MOVE = 22     # Continuous right move
        UP_SERIES_MOVE = 23        # Continuous up move
        DOWN_SERIES_MOVE = 24      # Continuous down move
        RANDOM = 25                # Random effect
        HT_LEFT_SERIES_MOVE = 26   # Head-to-tail continuous left
        HT_RIGHT_SERIES_MOVE = 27  # Head-to-tail continuous right
        HT_UP_SERIES_MOVE = 28     # Head-to-tail continuous up
        HT_DOWN_SERIES_MOVE = 29   # Head-to-tail continuous down
        FLICKEREFFECT = 30         # Flicker effect

        @classmethod
        def from_int(cls, value: int):
            for effect_type in cls:
                if effect_type.value == value:
                    return effect_type
            return cls.IMMEDIATE_SHOW  # Default for unknown values

    def __init__(self, *args, **kwargs):
        if len(args) == 1 and isinstance(args[0], Effect):
            # Copy constructor
            other = args[0]
            self._type = other._type
            self.speed = other.speed
            self.hold = other.hold
        else:
            # Handle different argument counts
            self._type = args[0] if len(args) >= 1 else self.EffectType.IMMEDIATE_SHOW
            self.speed = args[1] if len(args) >= 2 else 5
            self.hold = args[2] if len(args) >= 3 else 5000

    @property
    def type(self) -> EffectType:
        return self._type

    @type.setter
    def type(self, value: EffectType):
        self._type = value

    @property
    def type_value(self) -> int:
        return self._type.value

    @type_value.setter
    def type_value(self, value: int):
        self._type = self.EffectType.from_int(value)

    def get_speed(self) -> int:
        return self.speed

    def set_speed(self, speed: int):
        self.speed = speed

    def get_hold(self) -> int:
        return self.hold

    def set_hold(self, hold: int):
        self.hold = hold

    def to_dict(self) -> dict:
        return {
            "type": self.type_value,
            "speed": self.speed,
            "hold": self.hold
        }

    @classmethod
    def from_dict(cls, data: dict):
        effect = cls()
        if "type" in data:
            effect.type_value = data["type"]
        if "speed" in data:
            effect.speed = data["speed"]
        if "hold" in data:
            effect.hold = data["hold"]
        return effect

    def __repr__(self):
        return f"Effect(type={self._type.name}, speed={self.speed}, hold={self.hold})"